Sources/SwiftCompilerPlugin/PluginMacroExpansionContext.swift

```diff
@@ -0,0 +1,253 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift open source project
+//
+// Copyright (c) 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftDiagnostics
+import SwiftParser
+import SwiftOperators
+import SwiftSyntax
+import SwiftSyntaxMacros
+
+/// Manages known source code combined with their filename/fileID. This can be
+/// used to get line/column from a syntax node in the managed source code.
+class SourceManager {
+  class KnownSourceSyntax {
+    struct Location {
+      var offset: Int
+      var line: Int
+      var column: Int
+      var fileID: String
+      var fileName: String
+    }
+
+    let node: Syntax
+    let location: Location
+
+    init(node: Syntax, location: Location) {
+      self.node = node
+      self.location = location
+    }
+
+    /// Location converter to get line/column in the node.
+    lazy var locationConverter: SourceLocationConverter = .init(
+      file: self.location.fileName,
+      tree: self.node
+    )
+  }
+
+  struct SourcePosition {
+    var fileName: String
+    var utf8Offset: Int
+  }
+
+  struct SourceRange {
+    var fileName: String
+    var startUTF8Offset: Int
+    var endUTF8Offset: Int
+  }
+
+  /// Syntax added by `add(_:)` method. Keyed by the `id` of the node.
+  private var knownSourceSyntax: [Syntax.ID: KnownSourceSyntax] = [:]
+
+  /// Convert syntax information to a `Syntax` node. The location informations
+  /// are cached in the source manager to provide `location(of:)` et al.
+  func add(_ syntaxInfo: PluginMessage.Syntax) -> Syntax {
+
+    var node: Syntax
+    var parser = Parser(syntaxInfo.source)
+    switch syntaxInfo.kind {
+    case .declaration:
+      node = Syntax(DeclSyntax.parse(from: &parser))
+    case .statement:
+      node = Syntax(StmtSyntax.parse(from: &parser))
+    case .expression:
+      node = Syntax(ExprSyntax.parse(from: &parser))
+    case .type:
+      node = Syntax(TypeSyntax.parse(from: &parser))
+    case .pattern:
+      node = Syntax(PatternSyntax.parse(from: &parser))
+    case .attribute:
+      node = Syntax(AttributeSyntax.parse(from: &parser))
+    }
+    do {
+      node = try OperatorTable.standardOperators.foldAll(node)
+    } catch let error {
+      // TODO: Error handling.
```

Comment:
Just to make sure: You know that this is left over? If operator folding fails, it might be OK to just leave the unresolved nodes in the tree, i.e. just keep the `catch` clause empty.

Reply:
Added a no-op error handler to make it non-`throws`